cpp/src/strings/translate.cu

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/strings/translate.hpp>
+#include <strings/utilities.hpp>
+#include <strings/utilities.cuh>
+
+#include <algorithm>
+
+namespace cudf
+{
+namespace strings
+{
+namespace detail
+{
+
+using translate_table = thrust::pair<char_utf8,char_utf8>;
+
+namespace
+{
+
+/**
+ * @brief This is the translate functor for replacing individual characters
+ * in each string.
+ */
+struct translate_fn
+{
+    column_device_view const d_strings;
+    translate_table const* table;
+    size_type table_size;
+    int32_t const* d_offsets{};
+    char* d_chars{};
+
+    __device__ size_type operator()(size_type idx)
+    {
+        if( d_strings.is_null(idx) )
+            return 0;
+        string_view d_str = d_strings.element<string_view>(idx);
+        size_type bytes = d_str.size_bytes();
+        char* out_ptr = nullptr;
+        if( d_offsets )
+            out_ptr = d_chars + d_offsets[idx];
+        for( auto chr : d_str )
+        {
+            auto entry = thrust::find_if( thrust::seq, table, table+table_size,
+                [chr] __device__ ( auto te ) { return te.first==chr; } );
+            if( entry < table + table_size )
```

Comment:
Is there a better way to do this? We are doing a `<` comparison on a `InputIterator` and a `const*` to `thrust::pairs` with `size_type` added to it? Seems far from ideal.

Reply:
Changed this to use begin/end iterator instead.